File: endpoints/swift/Reliable.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Native pure-Swift reliable XRCE stream layer (DDS-XRCE 1.0 §8.4.10/§8.4.11),
// byte-identical to `crates/xrce` and the other endpoint SDKs (`endpoints/zig`,
// `endpoints/c`, …). Self-contained (only `Foundation`, for `Date`/`Thread`/
// `NSLock`/`DispatchSemaphore`) so it compiles as its own module — the same
// recipe the ping-pong test uses for `Zerodds.swift` — and never collides with
// a generated IDL module's own wire-core types (`Endianness`/`Writer`/`Reader`).
//
// Two roles:
//   ReliableSender   — submit / pendingHeartbeat / recvAckNack / getInFlight
//                       (history + retransmit), plus `AsyncWriter`: a
//                       lock-guarded producer queue drained by a dedicated
//                       `Thread` that owns the `ReliableSender` state
//                       exclusively and performs every syscall — the
//                       producer's `write` never touches the network.
//   ReliableReceiver — recvData / drainInOrder / pendingAckNack (reorder +
//                       de-duplication).
//
// RFC-1982 16-bit sequence numbers; sender window 16, receiver buffer 64,
// heartbeat 500 ms, payload <= 65535.

import Foundation

public let reliableSenderWindow = 16
public let reliableReceiverBuffer = 64
public let reliableMaxPayload = 65_535
public let reliableHeartbeatPeriodMs: Int64 = 500
public let reliableStreamId: UInt8 = 0x80

private let halfWindow: UInt16 = 0x8000

/// RFC-1982: is `a` strictly before `b`?
public func seqLt(_ a: UInt16, _ b: UInt16) -> Bool {
    let d = b &- a
    return d != 0 && d < halfWindow
}

/// RFC-1982: is `a` strictly after `b`?
public func seqGt(_ a: UInt16, _ b: UInt16) -> Bool { seqLt(b, a) }

public enum ReliableError: Error, Equatable {
    case payloadTooLarge
    case windowFull
    case bufferFull
}

public struct Heartbeat: Equatable {
    public var firstUnacked: UInt16
    public var lastUnacked: UInt16
    public var streamId: UInt8
    public init(firstUnacked: UInt16, lastUnacked: UInt16, streamId: UInt8) {
        self.firstUnacked = firstUnacked
        self.lastUnacked = lastUnacked
        self.streamId = streamId
    }
}

public struct AckNack: Equatable {
    public var firstUnacked: UInt16
    public var nackBitmap: UInt16
    public var streamId: UInt8
    public init(firstUnacked: UInt16, nackBitmap: UInt16, streamId: UInt8) {
        self.firstUnacked = firstUnacked
        self.nackBitmap = nackBitmap
        self.streamId = streamId
    }
}

// ------------------------------------------------------------------
// Wire framing — 8-byte header + body, all little-endian:
//   [0]=session [1]=stream [2..4]=seq LE [4]=submsg id [5]=flags
//   [6..8]=body-len LE [8..]=body
// WRITE_DATA keeps the sample's RFC-1982 seq in the header seq field
// ([2..4]) with stream byte 0x80 (reliable), matching the Rust peer's
// `reliable_unframe`. HEARTBEAT/ACKNACK use the byte-golden layout (session
// 0x80, stream NONE 0x00, msg-seq=1) — `golden_heartbeat_le.bin` /
// `golden_acknack_le.bin`; every peer's parser ignores the header stream/seq
// bytes and only checks the submessage id at offset 4, so this is wire-safe.
// ------------------------------------------------------------------

/// Reliable WRITE_DATA frame; the 2-byte stream-header seq carries the sample seq.
public func writeDataFrame(seq: UInt16, sample: [UInt8]) -> [UInt8] {
    let n = UInt16(sample.count)
    var out: [UInt8] = [
        0x80, reliableStreamId,
        UInt8(seq & 0xff), UInt8(seq >> 8),
        0x07, 0x03,
        UInt8(n & 0xff), UInt8(n >> 8),
    ]
    out.append(contentsOf: sample)
    return out
}

/// Parses a reliable WRITE_DATA frame into `(seq, sample)`.
public func parseWriteData(_ frame: [UInt8]) -> (seq: UInt16, sample: [UInt8])? {
    guard frame.count >= 8, frame[4] == 0x07 else { return nil }
    let seq = UInt16(frame[2]) | (UInt16(frame[3]) << 8)
    return (seq, Array(frame[8...]))
}

/// HEARTBEAT frame (golden layout, 13 bytes).
public func heartbeatFrame(_ hb: Heartbeat) -> [UInt8] {
    [
        0x80, 0x00, 0x01, 0x00,
        0x0B, 0x01, 0x05, 0x00,
        UInt8(hb.firstUnacked & 0xff), UInt8(hb.firstUnacked >> 8),
        UInt8(hb.lastUnacked & 0xff), UInt8(hb.lastUnacked >> 8),
        hb.streamId,
    ]
}

/// Parses a HEARTBEAT frame. Ignores header stream/seq — matches every peer.
public func parseHeartbeat(_ frame: [UInt8]) -> Heartbeat? {
    guard frame.count >= 13, frame[4] == 0x0B else { return nil }
    return Heartbeat(
        firstUnacked: UInt16(frame[8]) | (UInt16(frame[9]) << 8),
        lastUnacked: UInt16(frame[10]) | (UInt16(frame[11]) << 8),
        streamId: frame[12]
    )
}

/// ACKNACK frame (golden layout, 13 bytes).
public func acknackFrame(_ an: AckNack) -> [UInt8] {
    [
        0x80, 0x00, 0x01, 0x00,
        0x0A, 0x01, 0x05, 0x00,
        UInt8(an.firstUnacked & 0xff), UInt8(an.firstUnacked >> 8),
        UInt8(an.nackBitmap & 0xff), UInt8(an.nackBitmap >> 8),
        an.streamId,
    ]
}

/// Parses an ACKNACK frame. Ignores header stream/seq — matches every peer.
public func parseAckNack(_ frame: [UInt8]) -> AckNack? {
    guard frame.count >= 13, frame[4] == 0x0A else { return nil }
    return AckNack(
        firstUnacked: UInt16(frame[8]) | (UInt16(frame[9]) << 8),
        nackBitmap: UInt16(frame[10]) | (UInt16(frame[11]) << 8),
        streamId: frame[12]
    )
}

// ------------------------------------------------------------------
// Sender
// ------------------------------------------------------------------

/// Sender-side state machine: buffers in-flight samples (history for
/// retransmit), assigns monotonic RFC-1982 seqnrs, tracks heartbeat cadence,
/// prunes on ACKNACK. Not thread-safe by itself — `AsyncWriter` below owns one
/// instance exclusively on its drain thread.
public final class ReliableSender {
    private var nextSeq: UInt16 = 0
    private var inFlight: [UInt16: [UInt8]] = [:]
    private var lastHeartbeatMs: Int64 = -1

    public init() {}

    public var inFlightCount: Int { inFlight.count }

    public func getInFlight(_ seq: UInt16) -> [UInt8]? { inFlight[seq] }

    /// In-flight seqnrs, ascending — a deterministic retransmit order.
    public func inFlightSeqs() -> [UInt16] { inFlight.keys.sorted() }

    /// Buffers a new sample, returns its assigned seqnr.
    public func submit(_ payload: [UInt8]) throws -> UInt16 {
        if payload.count > reliableMaxPayload { throw ReliableError.payloadTooLarge }
        if inFlight.count >= reliableSenderWindow { throw ReliableError.windowFull }
        let seq = nextSeq
        inFlight[seq] = payload
        nextSeq = nextSeq &+ 1
        return seq
    }

    /// Emits a HEARTBEAT if the window is non-empty and the period elapsed.
    /// `nowMs` is a monotonic millisecond clock supplied by the caller.
    public func pendingHeartbeat(nowMs: Int64) -> Heartbeat? {
        guard !inFlight.isEmpty else { return nil }
        let due = lastHeartbeatMs < 0 || (nowMs - lastHeartbeatMs) >= reliableHeartbeatPeriodMs
        guard due else { return nil }
        lastHeartbeatMs = nowMs
        let keys = inFlight.keys
        guard let first = keys.min(), let last = keys.max() else { return nil }
        return Heartbeat(firstUnacked: first, lastUnacked: last, streamId: reliableStreamId)
    }

    /// Processes an ACKNACK: everything strictly before `base` is acked
    /// (RFC-1982); within `[base, base+16)` a clear bit is acked (dropped), a
    /// set bit stays in-flight (retransmit candidate).
    public func recvAckNack(_ an: AckNack) {
        let base = an.firstUnacked
        let acked = inFlight.keys.filter { seqLt($0, base) }
        for seq in acked { inFlight.removeValue(forKey: seq) }
        for i: UInt16 in 0..<16 {
            let seq = base &+ i
            let bit = (an.nackBitmap >> i) & 1
            if bit == 0 { inFlight.removeValue(forKey: seq) }
        }
    }
}

// ------------------------------------------------------------------
// Receiver
// ------------------------------------------------------------------

/// Receiver-side state machine: reorders + de-duplicates incoming samples,
/// delivers gap-free runs starting at `expectedSeq`, computes the
/// missing-slot ACKNACK bitmap. Not thread-safe — one instance per stream.
public final class ReliableReceiver {
    private var expectedSeqStore: UInt16 = 0
    private var received: [UInt16: [UInt8]] = [:]

    public init() {}

    public var expectedSeq: UInt16 { expectedSeqStore }
    public var outOfOrderCount: Int { received.count }

    /// Buffers an incoming sample (dedup + DoS bound).
    public func recvData(seq: UInt16, payload: [UInt8]) throws {
        if seqLt(seq, expectedSeqStore) { return } // duplicate before window -> drop
        if received[seq] != nil { return } // already buffered
        if received.count >= reliableReceiverBuffer { throw ReliableError.bufferFull }
        received[seq] = payload
    }

    /// Delivers all contiguous samples from `expectedSeq`, advances it.
    public func drainInOrder() -> [(seq: UInt16, payload: [UInt8])] {
        var out: [(seq: UInt16, payload: [UInt8])] = []
        while let payload = received.removeValue(forKey: expectedSeqStore) {
            out.append((expectedSeqStore, payload))
            expectedSeqStore = expectedSeqStore &+ 1
        }
        return out
    }

    /// Computes the ACKNACK marking missing slots in `[expected, expected+16)`.
    /// `hintLastSeen` (typically a HEARTBEAT's `lastUnacked`) bounds the
    /// window so slots beyond it are not falsely marked missing.
    public func pendingAckNack(hintLastSeen: UInt16?) -> AckNack {
        let base = expectedSeqStore
        var bitmap: UInt16 = 0
        for i: UInt16 in 0..<16 {
            let seq = base &+ i
            if let h = hintLastSeen, seqGt(seq, h) { continue }
            if received[seq] == nil { bitmap |= (1 << i) }
        }
        return AckNack(firstUnacked: base, nackBitmap: bitmap, streamId: reliableStreamId)
    }

    public func reset() {
        expectedSeqStore = 0
        received.removeAll()
    }
}

// ------------------------------------------------------------------
// Async-decoupled writer: the producer's `write` only appends to a
// lock-guarded bounded queue and returns (back-pressure via a `false`
// result, never a block); a dedicated drain `Thread` owns the
// `ReliableSender` state exclusively and performs every syscall (submit ->
// history -> WRITE_DATA send; periodic HEARTBEAT; on ACKNACK retransmit +
// prune). The producer never enters the kernel.
// ------------------------------------------------------------------

public final class AsyncWriter {
    public static let queueCapacity = 1024

    private let sender = ReliableSender()
    private var queue: [[UInt8]] = []
    private let lock = NSLock()
    private let stopped = DispatchSemaphore(value: 0)
    private var thread: Thread?
    private let sendFn: ([UInt8]) -> Void
    private let recvFn: () -> [UInt8]?
    private let clock = Date()

    /// `sendFn` transmits one frame (e.g. `sendto` on a UDP socket); `recvFn`
    /// returns the next available received frame (e.g. non-blocking
    /// `recvfrom`), or `nil` when nothing is pending.
    public init(sendFn: @escaping ([UInt8]) -> Void, recvFn: @escaping () -> [UInt8]?) {
        self.sendFn = sendFn
        self.recvFn = recvFn
    }

    /// Starts the drain thread. Call once before `write`.
    public func start() {
        let t = Thread { [weak self] in self?.drainLoop() }
        t.name = "zerodds-reliable-writer"
        thread = t
        t.start()
    }

    /// Enqueues a sample for reliable delivery. Returns `false`
    /// (backpressure) when the queue is full — never blocks, never touches
    /// the network.
    @discardableResult
    public func write(_ sample: [UInt8]) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard queue.count < Self.queueCapacity else { return false }
        queue.append(sample)
        return true
    }

    private func pop() -> [UInt8]? {
        lock.lock()
        defer { lock.unlock() }
        return queue.isEmpty ? nil : queue.removeFirst()
    }

    private func drainLoop() {
        while !Thread.current.isCancelled {
            var idle = true
            if let sample = pop() {
                idle = false
                do {
                    let seq = try sender.submit(sample)
                    sendFn(writeDataFrame(seq: seq, sample: sample))
                } catch {
                    // Window full: drain pending ACKNACKs to free a slot, retry next tick.
                    _ = drainAckNacks()
                    continue
                }
            }
            if drainAckNacks() { idle = false }
            let nowMs = Int64(Date().timeIntervalSince(clock) * 1000)
            if let hb = sender.pendingHeartbeat(nowMs: nowMs) {
                sendFn(heartbeatFrame(hb))
            }
            if idle { usleep(200) }
        }
        stopped.signal()
    }

    @discardableResult
    private func drainAckNacks() -> Bool {
        var any = false
        while let frame = recvFn() {
            guard let an = parseAckNack(frame) else { break }
            sender.recvAckNack(an)
            for seq in sender.inFlightSeqs() {
                if let data = sender.getInFlight(seq) {
                    sendFn(writeDataFrame(seq: seq, sample: data))
                }
            }
            any = true
        }
        return any
    }

    /// Stops the drain thread and waits for it to exit.
    public func close() {
        thread?.cancel()
        stopped.wait()
        thread = nil
    }
}

File: endpoints/csharp/Reliable.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Reliable XRCE stream (stream_id >= 128, DDS-XRCE spec Sec 8.4.10/8.4.11) for
// the pure-C# endpoint SDK (ADR 0013). Self-contained: the state machine is
// mirrored byte-for-byte from crates/xrce/src/reliable.rs (also
// endpoints/rust/src/reliable.rs), the WRITE_DATA/HEARTBEAT/ACKNACK wire codec
// is byte-identical to endpoints/c, endpoints/rust, and crates/xrce, and
// AsyncReliableWriter is the async-decoupled writer: a bounded
// System.Threading.Channels ring (SPSC) plus a dedicated drain Thread that owns
// the ReliableSender and the UDP socket. The producer's Enqueue is a
// non-blocking channel write -- it never performs I/O; the drain thread does all
// sends, the HEARTBEAT timer, and ACKNACK-driven retransmit.
//
// Wire (little-endian; byte-identical to golden_heartbeat_le.bin /
// golden_acknack_le.bin from zerodds-endpoint-golden): an 8-byte header
// [session, stream, seq_lo, seq_hi, submsg, flags, len_lo, len_hi] then body.
// WRITE_DATA id=0x07 flags=0x03 on stream 0x80 (header seq = the sample's
// RFC-1982 seqnr). HEARTBEAT id=0x0B / ACKNACK id=0x0A, flags=0x01, on stream
// 0x00 (control); body is 5 bytes, the target reliable stream id travels in the
// body's last byte.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Channels;

namespace ZeroDDS.Endpoint
{
    /// Wire constants and the frame codec.
    public static class ReliableWire
    {
        public const byte Session = 0x80;
        public const byte StreamReliable = 0x80;
        public const byte StreamNone = 0x00;
        public const byte SmWriteData = 0x07;
        public const byte SmAckNack = 0x0A;
        public const byte SmHeartbeat = 0x0B;
        public const byte WriteFlags = 0x03;
        public const byte CtrlFlags = 0x01;

        /// Sender window cap -- matches the 16-bit ACKNACK bitmap.
        public const int Window = 16;

        /// Receiver out-of-order buffer cap (DoS bound).
        public const int ReceiverBuffer = 64;

        /// Per-sample payload cap (u16 submessage length limit).
        public const int MaxPayload = 65_535;

        /// Heartbeat period (spec recommends 100 ms; 500 ms conservative without a
        /// Tx pacing layer underneath).
        public static readonly TimeSpan HeartbeatPeriod = TimeSpan.FromMilliseconds(500);

        /// RFC-1982 16-bit "a is strictly before b" (wrapping serial comparison).
        public static bool SeqLt(ushort a, ushort b) => (short)(a - b) < 0;

        /// RFC-1982 16-bit "a is strictly after b".
        public static bool SeqGt(ushort a, ushort b) => (short)(a - b) > 0;

        /// Builds a reliable WRITE_DATA frame; the header seq carries the sample's
        /// RFC-1982 seqnr.
        public static byte[] WriteFrame(ushort seq, byte[] sample)
        {
            var o = new byte[8 + sample.Length];
            o[0] = Session;
            o[1] = StreamReliable;
            o[2] = (byte)(seq & 0xFF);
            o[3] = (byte)(seq >> 8);
            o[4] = SmWriteData;
            o[5] = WriteFlags;
            ushort n = (ushort)sample.Length;
            o[6] = (byte)(n & 0xFF);
            o[7] = (byte)(n >> 8);
            Array.Copy(sample, 0, o, 8, sample.Length);
            return o;
        }

        /// Parses a reliable WRITE_DATA frame into (seq, sample).
        public static bool TryUnframeWrite(ReadOnlySpan<byte> frame, out ushort seq, out byte[] sample)
        {
            if (frame.Length < 8 || frame[4] != SmWriteData)
            {
                seq = 0;
                sample = Array.Empty<byte>();
                return false;
            }
            seq = (ushort)(frame[2] | (frame[3] << 8));
            sample = frame[8..].ToArray();
            return true;
        }

        private static byte[] CtrlHeader(byte submsg, ushort msgSeq) => new byte[]
        {
            Session, StreamNone,
            (byte)(msgSeq & 0xFF), (byte)(msgSeq >> 8),
            submsg, CtrlFlags, 5, 0,
        };

        /// Builds a HEARTBEAT control frame. Byte-identical to golden_heartbeat_le.bin
        /// when hb = (1, 3, 0x80) and msgSeq = 1.
        public static byte[] HeartbeatFrame(Heartbeat hb, ushort msgSeq)
        {
            var f = new List<byte>(13);
            f.AddRange(CtrlHeader(SmHeartbeat, msgSeq));
            f.Add((byte)(hb.First & 0xFF));
            f.Add((byte)((hb.First >> 8) & 0xFF));
            f.Add((byte)(hb.Last & 0xFF));
            f.Add((byte)((hb.Last >> 8) & 0xFF));
            f.Add(hb.StreamId);
            return f.ToArray();
        }

        /// Parses a HEARTBEAT frame (keys on the submessage id; header otherwise ignored).
        public static bool TryParseHeartbeat(ReadOnlySpan<byte> frame, out Heartbeat hb)
        {
            if (frame.Length < 13 || frame[4] != SmHeartbeat)
            {
                hb = default;
                return false;
            }
            hb = new Heartbeat(
                (short)(frame[8] | (frame[9] << 8)),
                (short)(frame[10] | (frame[11] << 8)),
                frame[12]);
            return true;
        }

        /// Builds an ACKNACK control frame. Byte-identical to golden_acknack_le.bin
        /// when ack = (1, bitmap=0, 0x80) and msgSeq = 1.
        public static byte[] AckNackFrame(AckNack ack, ushort msgSeq)
        {
            var f = new List<byte>(13);
            f.AddRange(CtrlHeader(SmAckNack, msgSeq));
            f.Add((byte)(ack.FirstUnacked & 0xFF));
            f.Add((byte)((ack.FirstUnacked >> 8) & 0xFF));
            f.Add(ack.NackLo);
            f.Add(ack.NackHi);
            f.Add(ack.StreamId);
            return f.ToArray();
        }

        /// Parses an ACKNACK frame (keys on the submessage id; header otherwise ignored).
        public static bool TryParseAckNack(ReadOnlySpan<byte> frame, out AckNack ack)
        {
            if (frame.Length < 13 || frame[4] != SmAckNack)
            {
                ack = default;
                return false;
            }
            ack = new AckNack(
                (short)(frame[8] | (frame[9] << 8)),
                frame[10],
                frame[11],
                frame[12]);
            return true;
        }
    }

    /// HEARTBEAT body.
    public readonly struct Heartbeat
    {
        public readonly short First;
        public readonly short Last;
        public readonly byte StreamId;

        public Heartbeat(short first, short last, byte streamId)
        {
            First = first;
            Last = last;
            StreamId = streamId;
        }
    }

    /// ACKNACK body. The 16-bit NACK bitmap travels as two little-endian bytes
    /// (bit i set => seqnr FirstUnacked+i still missing).
    public readonly struct AckNack
    {
        public readonly short FirstUnacked;
        public readonly byte NackLo;
        public readonly byte NackHi;
        public readonly byte StreamId;

        public AckNack(short firstUnacked, byte nackLo, byte nackHi, byte streamId)
        {
            FirstUnacked = firstUnacked;
            NackLo = nackLo;
            NackHi = nackHi;
            StreamId = streamId;
        }

        public ushort Bitmap => (ushort)(NackLo | (NackHi << 8));
    }

    public enum SubmitStatus { Ok, PayloadTooLarge, WindowFull }

    public enum RecvStatus { Ok, BufferFull }

    /// Reliable sender: assigns seqnrs, holds in-flight samples until
    /// acknowledged, emits HEARTBEATs, and clears/keeps samples on ACKNACK.
    /// Mirrors the sender half of crates/xrce/src/reliable.rs ReliableStreamState.
    public sealed class ReliableSender
    {
        private ushort _nextSeq;
        private readonly SortedDictionary<ushort, byte[]> _inFlight = new();
        private DateTime? _lastHeartbeat;

        /// In-flight (unacknowledged) sample count.
        public int InFlightCount => _inFlight.Count;

        /// Submits a sample, assigning it the next seqnr.
        public SubmitStatus Submit(byte[] payload, out ushort seq)
        {
            seq = 0;
            if (payload.Length > ReliableWire.MaxPayload) return SubmitStatus.PayloadTooLarge;
            if (_inFlight.Count >= ReliableWire.Window) return SubmitStatus.WindowFull;
            seq = _nextSeq;
            _inFlight[seq] = payload;
            _nextSeq = (ushort)(_nextSeq + 1);
            return SubmitStatus.Ok;
        }

        /// The in-flight payload for `seq`, if still unacknowledged (for retransmit).
        public byte[]? GetInFlight(ushort seq) => _inFlight.TryGetValue(seq, out var p) ? p : null;

        /// All in-flight seqnrs, ascending (for retransmit against a NACK bitmap).
        public IReadOnlyList<ushort> InFlightSeqs()
        {
            var seqs = new List<ushort>(_inFlight.Count);
            seqs.AddRange(_inFlight.Keys);
            return seqs;
        }

        /// Returns a HEARTBEAT if the period elapsed and samples are in flight.
        public Heartbeat? PendingHeartbeat(DateTime now)
        {
            if (_inFlight.Count == 0) return null;
            bool due = _lastHeartbeat is null || (now - _lastHeartbeat.Value) >= ReliableWire.HeartbeatPeriod;
            if (!due) return null;
            _lastHeartbeat = now;
            // RFC-1982 window base (oldest unacked) + end (newest unacked); NOT the
            // numeric SortedDictionary first/last key, which is wrong across a
            // 16-bit wrap (window 0xFFFE,0xFFFF,0x0000,0x0001 -> base 0xFFFE / end
            // 0x0001, not 0x0000 / 0xFFFF). Mirrors window_base /
            // serial_max_in_flight in crates/xrce/src/reliable.rs.
            ushort first = 0, last = 0;
            bool any = false;
            foreach (var k in _inFlight.Keys)
            {
                if (!any)
                {
                    first = k;
                    last = k;
                    any = true;
                }
                else
                {
                    if (ReliableWire.SeqLt(k, first)) first = k;
                    if (ReliableWire.SeqGt(k, last)) last = k;
                }
            }
            return new Heartbeat((short)first, (short)last, ReliableWire.StreamReliable);
        }

        /// Processes an ACKNACK: drops everything strictly before `FirstUnacked`
        /// and every clear-bit slot in [FirstUnacked, FirstUnacked+16); keeps
        /// set-bit (still-missing) slots for retransmit.
        public void RecvAckNack(AckNack ack)
        {
            ushort baseSeq = (ushort)ack.FirstUnacked;
            ushort bitmap = ack.Bitmap;
            var toDrop = new List<ushort>();
            foreach (var k in _inFlight.Keys)
            {
                if (ReliableWire.SeqLt(k, baseSeq)) toDrop.Add(k);
            }
            foreach (var k in toDrop) _inFlight.Remove(k);
            for (int i = 0; i < 16; i++)
            {
                ushort seq = (ushort)(baseSeq + i);
                if (((bitmap >> i) & 1) == 0) _inFlight.Remove(seq);
            }
        }

        /// Re-arms the heartbeat clock so the next tick fires immediately
        /// (used by the drain loop while it waits out a close-drain).
        internal void ResetHeartbeatClock() => _lastHeartbeat = null;
    }

    /// Reliable receiver: buffers out-of-order samples, delivers them
    /// contiguously, and reports the missing slots via ACKNACK.
    public sealed class ReliableReceiver
    {
        private ushort _expected;
        private readonly SortedDictionary<ushort, byte[]> _received = new();

        /// Next expected incoming seqnr.
        public ushort Expected => _expected;

        /// Buffered out-of-order sample count.
        public int OutOfOrderCount => _received.Count;

        /// Accepts a sample; drops duplicates before `Expected` or already buffered.
        public RecvStatus RecvData(ushort seq, byte[] payload)
        {
            if (ReliableWire.SeqLt(seq, _expected)) return RecvStatus.Ok; // duplicate, drop
            if (_received.ContainsKey(seq)) return RecvStatus.Ok; // already buffered
            if (_received.Count >= ReliableWire.ReceiverBuffer) return RecvStatus.BufferFull;
            _received[seq] = payload;
            return RecvStatus.Ok;
        }

        /// Delivers all contiguous samples from `Expected`, advancing it.
        public List<(ushort Seq, byte[] Payload)> DrainInOrder()
        {
            var outp = new List<(ushort, byte[])>();
            while (_received.TryGetValue(_expected, out var payload))
            {
                outp.Add((_expected, payload));
                _received.Remove(_expected);
                _expected = (ushort)(_expected + 1);
            }
            return outp;
        }

        /// Computes the ACKNACK marking the missing slots in [Expected, Expected+16).
        /// A clear bit unambiguously means "received" when `hint` is omitted; with
        /// a `hint` (e.g. the HEARTBEAT's last-in-flight seqnr), slots strictly
        /// after it are left un-marked (not yet sent by the peer).
        public AckNack PendingAckNack(ushort? hint = null)
        {
            ushort bitmap = 0;
            for (int i = 0; i < 16; i++)
            {
                ushort seq = (ushort)(_expected + i);
                if (hint.HasValue && ReliableWire.SeqGt(seq, hint.Value)) continue;
                if (!_received.ContainsKey(seq)) bitmap |= (ushort)(1 << i);
            }
            return new AckNack((short)_expected, (byte)(bitmap & 0xFF), (byte)(bitmap >> 8), ReliableWire.StreamReliable);
        }

        /// Clears all receiver state.
        public void Reset()
        {
            _expected = 0;
            _received.Clear();
        }
    }

    /// Shared closed-flag between a `ReliableWriterHandle` and its
    /// `AsyncReliableWriter`. Tracked independently of the channel's own
    /// completion (which only completes once the channel has also been fully
    /// *drained* -- exactly the condition the drain thread needs to detect
    /// separately, since a stalled window can leave items unread indefinitely).
    internal sealed class WriterCloseState
    {
        private volatile bool _closed;
        public bool Closed => _closed;
        public void SetClosed() => _closed = true;
    }

    /// Producer-side handle to the async-decoupled writer. `Enqueue` is a
    /// non-blocking channel write -- it never performs I/O; the drain thread does
    /// all sends. `Close` signals the drain thread that no more samples are
    /// coming, once the send window has drained it exits.
    public sealed class ReliableWriterHandle
    {
        private readonly ChannelWriter<byte[]> _writer;
        private readonly WriterCloseState _state;

        internal ReliableWriterHandle(ChannelWriter<byte[]> writer, WriterCloseState state)
        {
            _writer = writer;
            _state = state;
        }

        /// Non-blocking enqueue. Returns false (backpressure) if the ring is full;
        /// the caller should retry (e.g. after a `Thread.Yield()`), the same
        /// contract as the Rust/D/C SPSC-ring writers.
        public bool Enqueue(byte[] sample) => _writer.TryWrite(sample);

        /// Signals the drain thread that no more samples will be enqueued.
        public void Close()
        {
            _state.SetClosed();
            _writer.TryComplete();
        }
    }

    /// The async-decoupled reliable writer: a bounded, single-writer/single-reader
    /// System.Threading.Channels ring plus a dedicated drain Thread that owns the
    /// ReliableSender and the UDP socket. Mirrors
    /// endpoints/rust/src/reliable.rs::AsyncReliableWriter.
    public sealed class AsyncReliableWriter : IDisposable
    {
        private readonly Channel<byte[]> _channel;
        private readonly System.Net.Sockets.Socket _sock;
        private readonly Thread _drain;
        private readonly WriterCloseState _closeState = new();
        private bool _shutdownCalled;

        /// A handle the producer(s) hold; Enqueue never touches the kernel.
        public ReliableWriterHandle Handle { get; }

        private AsyncReliableWriter(System.Net.Sockets.Socket sock, int capacity)
        {
            _sock = sock;
            _channel = Channel.CreateBounded<byte[]>(new BoundedChannelOptions(capacity)
            {
                SingleWriter = true,
                SingleReader = true,
                FullMode = BoundedChannelFullMode.Wait,
            });
            Handle = new ReliableWriterHandle(_channel.Writer, _closeState);
            _drain = new Thread(DrainLoop) { IsBackground = true, Name = "zerodds-reliable-drain" };
            _drain.Start();
        }

        /// Spawns the drain thread. `sock` must already target the peer (connected
        /// UDP socket); the drain thread owns it exclusively from here on.
        public static AsyncReliableWriter Start(System.Net.Sockets.Socket sock, int capacity = 1024) =>
            new(sock, capacity);

        private void DrainLoop()
        {
            _sock.ReceiveTimeout = 5;
            var sender = new ReliableSender();
            ushort ctlSeq = 1;
            var buf = new byte[2048];
            DateTime? closeDeadline = null;

            while (true)
            {
                // 1) Drain the channel into the sender window; send fresh WRITE_DATA.
                while (sender.InFlightCount < ReliableWire.Window && _channel.Reader.TryRead(out var sample))
                {
                    if (sender.Submit(sample, out var seq) == SubmitStatus.Ok)
                    {
                        TrySend(ReliableWire.WriteFrame(seq, sample));
                    }
                }

                // 2) HEARTBEAT when due.
                var hb = sender.PendingHeartbeat(DateTime.UtcNow);
                if (hb.HasValue)
                {
                    TrySend(ReliableWire.HeartbeatFrame(hb.Value, ctlSeq));
                    ctlSeq++;
                }

                // 3) Drain incoming ACKNACKs -> retransmit the still-missing samples.
                while (TryReceive(buf, out int n) && n > 0)
                {
                    if (ReliableWire.TryParseAckNack(buf.AsSpan(0, n), out var ack))
                    {
                        sender.RecvAckNack(ack);
                        ushort baseSeq = (ushort)ack.FirstUnacked;
                        ushort bitmap = ack.Bitmap;
                        for (int i = 0; i < 16; i++)
                        {
                            if (((bitmap >> i) & 1) != 1) continue;
                            ushort seq = (ushort)(baseSeq + i);
                            var p = sender.GetInFlight(seq);
                            if (p != null) TrySend(ReliableWire.WriteFrame(seq, p));
                        }
                    }
                }

                // 4) Exit once closed AND the window has drained (ring empty, every
                //    in-flight sample acknowledged) -- or, as a safety valve, once
                //    the post-close drain deadline elapses (a vanished peer -- or a
                //    permanently-full window with nobody ACKing -- must never wedge
                //    Shutdown() forever).
                if (_closeState.Closed)
                {
                    closeDeadline ??= DateTime.UtcNow + TimeSpan.FromSeconds(5);
                    bool ringEmpty = !_channel.Reader.TryPeek(out _);
                    if (sender.InFlightCount == 0 && ringEmpty) break;
                    if (DateTime.UtcNow > closeDeadline) break; // best-effort give-up
                    // Keep heartbeating so the peer keeps ACKing until the window empties.
                    sender.ResetHeartbeatClock();
                }
            }
        }

        private void TrySend(byte[] frame)
        {
            try { _sock.Send(frame); }
            catch (System.Net.Sockets.SocketException) { /* best-effort */ }
        }

        private bool TryReceive(byte[] buf, out int n)
        {
            try
            {
                n = _sock.Receive(buf);
                return true;
            }
            catch (System.Net.Sockets.SocketException)
            {
                n = 0;
                return false;
            }
        }

        /// Closes the writer and blocks until the drain thread finishes (the
        /// window has drained, or the best-effort deadline elapsed). Idempotent.
        public void Shutdown()
        {
            if (_shutdownCalled)
            {
                _drain.Join();
                return;
            }
            _shutdownCalled = true;
            Handle.Close();
            _drain.Join();
        }

        public void Dispose() => Shutdown();
    }
}
